proposals/0000-replacing-cabal-custom-build.md

```diff
@@ -0,0 +1,1775 @@
+     -- has changed.
+  , monitoredDirs :: ![ Location ]
+     -- ^ Monitored directories; if the contents of these directories change,
+     -- the rule is considered out-of-date.
+  }
+
+newtype Action =
+  Action
+    { action
+      :: [ ResolvedLocation ]
+           -- ^ locations at which the __dependencies__ of this action
+           -- were found
+      -> [ ResolvedLocation ]
+           -- ^ locations in which the ___results__ of this action
+           -- should be put
+      -> IO () }
+```
+
+The specific monadic type of rules, namely
+
+```haskell
+env -> ActionsM ( IO [Rule] )
+```
+
+is meant to address concerns surrounding generation of `ActionId`s, as is
+explained in [§ Identifiers for fine-grained rules](#identifiers-for-fine-grained-rules).
+In practice, we can think of the rules as being specified by a Haskell
+function with the following type:
+
+```haskell
+env -> ( Map ActionId Action, IO [Rule] )
+```
+
+### Dependency structure
+
+Rule dependencies and results are declared using the following datatypes:
+
+```haskell
+data Dependency
+  -- | Declare a dependency on a file from the current project that should
+  -- be found by looking at project search paths.
+  --
+  -- This file might exist already, or it might be the output of another rule.
+  = ProjectSearchDirFile
+      !Location
+        -- ^ where to go looking for the file
+      !FilePath
+        -- ^ path of the file, relative to a Cabal search directory
+type Result = ( Location, FilePath )
+data Location
+  -- | A source file:
+  --
+  -- - for a rule dependency, we will go looking for it in
+  --   the source directories and in autogen modules directories;
+  -- - for a rule output, the file should be put in an autogen module directory.
+  = SrcFile
+  -- | A build-file, that belongs to some build directory.
+  | BuildFile
+  -- | A temporary file.
+  | TmpFile
+```
+
+To illustrate, a rule could declare a dependency on `Parser.y` (using the
+`dependencies` field). The build-system will look through appropriate
+search paths to resolve this dependency, and pass the location in which the file
+was found on disk as one of the elements of the list passed as the first argument
+to the `action` stored in the `Action` that the rule refers to through its
+`ActionId`. (Although see [§ No searching in fine-grained rules](#no-searching-in-fine-grained-rules)
+for an alternative in which searching is performed when computing the rules
+themselves.)
+
+Note that we do not allow a rule to directly depend on another rule, as this
+can easily introduce bugs. For example, suppose that `Rule {ruleId = 1}`
+outputs `A.y` and `Rule {ruleId = 2}` depends on `A.y` in order to produce
+`A.hs`. It is much more direct and robust for rule 2 to directly declare its
+dependency on `A.y`, rather than on (the output of) rule 1, as the latter is
+prone to breakage if one refactors the code and changes which rule generates
+`A.y`.
```

Comment:
I don't see how this can be true if one gets the location of `A.y` from one's dependency on the output of a rule. Then if the rule stops producing `A.y`, you won't get passed a location that contains `A.y` and so you find out.

Generally I think the reverse - the rule structure is fundamental, and files are just one kind of output from rules.

Reply:
I think we agree here. What this paragraph was trying to say is that it is better for dependencies to be described by paths rather than directly referring to other rules. For example

```haskell
  registerRule $ Rule { results = [ "A.y" ] }
  registerRule $ Rule { deps = [ "A.y" ], result = [ "A.hs" ], action = alex }
```

is more robust than:

```haskell
  rule1 <- registerRule $ Rule { results = [ "A.y" ] }
  rule2 <- registerRule $ Rule { deps = [ rule1 ], result = [ "A.hs" ], action = alex }
```

because the latter will break if you change which rule generates `A.y`. The latter style requires the user to perform dependency analysis themselves, whereas in the first example it is the build system which figures out the dependency.

Perhaps you are saying that the low-level API could use this style (in which rules directly refer to other rules), and provide an interface on top that would work as the current proposal describes? I'm not really sure what that would buy us, especially given that the IPC requirement means that it would be difficult to allow users to declare their own datatypes for use as `RuleId`s.

Reply:
I just don't find this at all convincing. I've worked with systems that work both ways and I find the version that works with rules much clearer and easier to understand. With the other version it is IMO harder to maintain because there is magic in figuring out what rule produces `A.y` when you look at the second rule definition. You can't just see what rule it is, you have to go through all the rules and look to see which ones produce `A.y`. 

> because the latter will break if you change which rule generates A.y

Can we have a realistic example? I just don't see how this is going to happen in a way that's bothersome. Yes, if you change a rule you may have to go and fix the things that depend on it... that doesn't seem at all surprising?

> The latter style requires the user to perform dependency analysis themselves, whereas in the first example it is the build system which figures out the dependency.

What? It requires the user to know which rule produces `A.y`, yes, but that's not doing dependency analysis, that's identifying the _direct_ dependencies.

Reply:
Perhaps a more convincing example builds on the compositionality of the
rules. For example, a given preprocessor (e.g. c2hs) could be implemented
as built-in rules in Cabal, or as a separate library. If I then want
additional rules on top, I should not need to know the internal
implementation details of the preprocessor rules. If I'm required to refer
to rules directly, I would have to inspect the monadic state of the rules
computation for the preprocessor to find which rule outputs the file I
might want to depend on, instead of directly declaring the dependency on
the file. It seems correct to me that the build system, which has a global
view of all rules, should resolve these dependencies, as opposed to asking
the rules author to do so; the latter feels inherently non-compositional to
me.